packages/babel-parser/src/plugins/typescript/index.js

```diff
@@ -3011,4 +3011,15 @@ export default (superClass: Class<Parser>): Class<Parser> =>
       }
       return method;
     }
+
+    shouldParseAsDts(): boolean {
+      return !!this.getPluginOption("typescript", "dts");
+    }
+
+    parse() {
+      if (this.shouldParseAsDts()) {
```

Comment:
Doing so all files would be parsed as declaration though, maybe we should check the filename also?

Reply:
The parser ignores filenames (e.g. it doesn't check if a file has a `.ts` extension). Only files for which the `dts: true` option is enabled will be parsed as declaration, and it's up to the `@babel/parser` consumer to decide when to enable it.